### PR TITLE
Update self contained/ runtime command line options for build and publish

### PR DESCRIPTION
**Pull request by sfoslund**
Fixes 4th and 5th bullets of https://github.com/dotnet/sdk/issues/18832

## Description
This change adds `--self-contained` and `--no-self-contained` options to the `dotnet build` command to achieve parity with `dotnet publish`. It also adds a warning for .NET 6+ apps when `--runtime` is specified on the CLI without `--self-contained` or `--no-self-contained`.
These changes are all additive and do not impact the existing CLI. Note that this doesn't change functionality, as users can achieve the same result by specifying self contained as a MSBuild property on the command line. This change will make it more straight forward to achieve the same result. 

## Regression?
No

## Risk
Low, this is a new feature and won't effect existing CLI. 

## Verification
[X] Manual (required)
[x] Automated

### CODE REVIEW

File: src/Cli/dotnet/CommonLocalizableStrings.resx
Comment (richlander) [inline src/Cli/dotnet/CommonLocalizableStrings.resx:681]: Can you call this `FrameworkDependent` instead of `NoSelfContained`? Framework dependent is the product default so it seems odd to pivot on self-contained. I know this is just a property name, but seems worth some consideration.

File: src/Cli/dotnet/CommonLocalizableStrings.resx
Comment (richlander) [inline src/Cli/dotnet/CommonLocalizableStrings.resx:682]:
```suggestion
    <value>Publish your application as a framework dependent application without the .NET runtime. A compatible .NET runtime must be installed to run your application.</value>
```

File: src/Cli/dotnet/CommonLocalizableStrings.resx
Comment (richlander) [inline src/Cli/dotnet/CommonLocalizableStrings.resx:687]: This property has the opposite polarity it should have, I believe. I think it should be: `SelfContainedOptionShouldBeUsedWithRuntime`.

File: src/Cli/dotnet/CommonLocalizableStrings.resx
Comment (richlander) [inline src/Cli/dotnet/CommonLocalizableStrings.resx:688]:
```suggestion
    <value>Warning: The  '--self-contained' or '--no-self-contained' option should be used with '-r|--runtime'.</value>
```

Comment (richlander) [reply src/Cli/dotnet/CommonLocalizableStrings.resx:688]: Similar problem.

**Comment by sfoslund** [comment]
@dsplaisted can I get a quick review here?

File: src/Cli/dotnet/CommonOptions.cs
Comment (dsplaisted) [inline src/Cli/dotnet/CommonOptions.cs:107]: Do we really want to support both `--no-self-contained` and `--self-contained false`?

@KathleenDollard

File: src/Cli/dotnet/CommonOptions.cs
Comment (dsplaisted) [inline src/Cli/dotnet/CommonOptions.cs:135]: Should we be evaluating the project in the CLI here?  What kind of a perf impact will that have?

@rainersigwald

Perhaps we should try to generate this warning from the project build itself.  That would mean it would be a NETSDK warning message with an error code, and could be ignored, which would also be nice.

Comment (sfoslund) [reply src/Cli/dotnet/CommonOptions.cs:135]: I added the error here because we wouldn't be able to detect when these properties were set on the command line vs by targets, but I suppose I could add an extra property with the CLI options to mark that it was set on the cli then move the error into targets. Do you want me to update with that approach?

Comment (rainersigwald) [reply src/Cli/dotnet/CommonOptions.cs:135]: Evaluation is expensive enough that it seems like a good idea to push this into targets (but I don't have all the context on this particular thing fresh in my mind).

Comment (dsplaisted) [reply src/Cli/dotnet/CommonOptions.cs:135]: Sounds like it would be a good idea to move it into the targets.  Though I think it's likely to generate the warning for each project built, as the properties we pass from the CLI are global properties which by default flow to referenced projects.

Comment (sfoslund) [reply src/Cli/dotnet/CommonOptions.cs]: Updated, moved to targets. @dsplaisted do you mind re-reviewing?

File: src/Tasks/Common/Resources/Strings.resx
Comment (dsplaisted) [inline src/Tasks/Common/Resources/Strings.resx:819]: @richlander @KathleenDollard To review the warning text here.  Perhaps something like the following would be clearer:

> When the target runtime is specified with '--runtime' or '-r', the deployment type should be specified with '--self-contained' or '--no-self-contained'.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets
Comment (dsplaisted) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Sdk.targets:1083]: For a lot of similar checks, we do `BeforeTargets="_CheckForInvalidConfigurationAndPlatform"`.  Is there any reason this is different?

File: src/Tests/dotnet-build.Tests/GivenDotnetBuildBuildsCsproj.cs
Comment (dsplaisted) [inline src/Tests/dotnet-build.Tests/GivenDotnetBuildBuildsCsproj.cs:197]: This needs to target .NET 6, or it won't generate the warning anyway, right?

**Review by dsplaisted — COMMENTED** [review]
The code changes look good.

Personally I don't think we should have a `--no-self-contained` option.  I think we should instead just use `--self-contained false`.  I'd like @KathleenDollard to weigh in on this, and am happy to defer to whatever she thinks is best here.

I also believe that we will now repeat this warning for each (transitively) referenced project.  Is that OK, or should we try to make the warning just show up once?

**Comment by sfoslund** [comment]
> Personally I don't think we should have a --no-self-contained option

For what its worth, `--no-self-contained is already an option for publish, so this PR just adds it to build for parity.

File: src/Tasks/Common/Resources/Strings.resx
Comment (sfoslund) [reply src/Tasks/Common/Resources/Strings.resx:819]: @richlander suggested this string in the previous PR reviews.

Comment (KathleenDollard) [reply src/Tasks/Common/Resources/Strings.resx:819]: That seems a bit long, and I like to reduce wrapping where possible. Maybe: 

```suggestion
    <value>NETSDK1178: The  '--self-contained' or '--no-self-contained' option should be used when  '--runtime' is specified.</value>
```

Daniel, is there a reason you want to lead with the conditional?

**Review by KathleenDollard — COMMENTED** [review]
And I had an old pending comment as well on this 

I'd like Rich's call on the string in question. I can live with any of the options as I think they are all clear.

File: src/Tasks/Common/Resources/Strings.resx
Comment (dsplaisted) [reply src/Tasks/Common/Resources/Strings.resx:819]: @KathleenDollard It sounded a bit better to me, but it could certainly be reversed.

File: src/Cli/dotnet/CommonLocalizableStrings.resx
Comment (sfoslund) [reply src/Cli/dotnet/CommonLocalizableStrings.resx]: This is not a new string, it's just showing up in the diff because it was moved to a different resx file. This is the existing help text.

File: src/Cli/dotnet/CommonLocalizableStrings.resx
Comment (sfoslund) [reply src/Cli/dotnet/CommonLocalizableStrings.resx]: This string has been removed in subsequent updates anyways, resolving this...

File: src/Tasks/Common/Resources/Strings.resx
Comment (richlander) [reply src/Tasks/Common/Resources/Strings.resx]: I like shorter version better.

How about this one?

> One of '--self-contained' or '--no-self-contained' options are required when '--runtime' is used.

**Comment by sfoslund** [comment]
@richlander @dsplaisted @KathleenDollard I've updated based on your feedback, thanks! We're hoping to get this in for preview 7 so can I please get another review?

File: src/Tasks/Common/Resources/Strings.resx
Comment (KathleenDollard) [reply src/Tasks/Common/Resources/Strings.resx]: I hate to keep this going, but...

> Either '--self-contained' or '--no-self-contained' is required when '--runtime' is used.

Aligning "One of" and "are" grammatically seems hard ;-)

File: src/Cli/dotnet/CommonOptions.cs
Comment (dsplaisted) [inline src/Cli/dotnet/CommonOptions.cs:118]: Seems like we don't need the last two arguments anymore.

**Review by dsplaisted — APPROVED** [review]

**Pull request merged**
